sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala

```diff
@@ -161,22 +164,70 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
             if (lazyPruningEnabled) {
               index
             } else {
               index.filterPartitions(Nil)  // materialize all the partitions in memory
             }
           }
           val partitionSchemaColumnNames = partitionSchema.map(_.name.toLowerCase).toSet
-          val dataSchema =
-            StructType(metastoreSchema
-              .filterNot(field => partitionSchemaColumnNames.contains(field.name.toLowerCase)))
+          val filteredMetastoreSchema = StructType(metastoreSchema
+            .filterNot(field => partitionSchemaColumnNames.contains(field.name.toLowerCase)))
+
+          // Infer a case-sensitive schema when the metastore doesn't return one, if configured.
+          val inferenceMode = sparkSession.sessionState.conf.caseSensitiveInferenceMode
+          val inferredSchema = if (!catalogTable.schemaPreservesCase &&
+              inferenceMode != HiveCaseSensitiveInferenceMode.NEVER_INFER) {
+            logInfo(s"Inferring case-sensitive schema for table $tableIdentifier")
+            val fileStatuses = fileIndex.listFiles(Nil).flatMap(_.files)
+            val inferred = defaultSource.inferSchema(sparkSession, options, fileStatuses)
+            if (fileType.equals("parquet")) {
+              inferred.map(ParquetFileFormat.mergeMetastoreParquetSchema(metastoreSchema, _))
+            } else {
+              inferred
+            }
+          } else {
+            None
+          }
+
+          // If configured, save the inferred case-sensitive schema to the table properties and
+          // fetch the updated CatalogTable record for use in the LogicalRelation.
+          val updatedCatalogTable = if (!catalogTable.schemaPreservesCase &&
+              inferenceMode == HiveCaseSensitiveInferenceMode.INFER_AND_SAVE) {
+            inferredSchema.flatMap { schema =>
+              logInfo(s"Saving case-sensitive schema for table $tableIdentifier to table " +
+                "properties")
+              val updatedTable = catalogTable.copy(schema = schema)
+              try {
+                val catalog = sparkSession.sharedState.externalCatalog
+                catalog.alterTable(updatedTable)
+                Option(catalog.getTable(updatedTable.identifier.database.get,
+                  updatedTable.identifier.table))
+              } catch {
+                case NonFatal(ex) =>
+                  logError(s"Error saving case-sensitive schema for table $tableIdentifier: $ex")
+                  None
+              }
+            }
+          } else {
+            None
+          }
+
+          val dataSchema = if (!catalogTable.schemaPreservesCase) {
+            inferredSchema.getOrElse {
+              logWarning(s"Unable to infer schema for table $tableIdentifier from file format " +
+                s"$defaultSource; using metastore schema.")
+              filteredMetastoreSchema
+            }
+          } else {
+            filteredMetastoreSchema
+          }
```

Comment:
This function is getting too long. How hard would it be to pull the schema inference stuff into a helper function?